destination.go

```diff
@@ -133,6 +152,133 @@ func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, err
 	return len(records), nil
 }
 
+func (d *Destination) sendRecordWithRetries(ctx context.Context, record *opencdcv1.Record) error {
+	d.streamMutex.Lock()
+	if d.stream == nil {
+		// stream would be nil in the case of a failed reconnection
+		err := <-d.errCh
+		return err
+	}
+	err := d.stream.Send(record)
+	if err == io.EOF || status.Code(err) == codes.Unavailable {
+		d.streamMutex.Unlock()
+		err := d.waitForReadyState(ctx, d.conn.GetState())
+		if err != nil {
+			return err
+		}
+		return d.sendRecordWithRetries(ctx, record)
+	}
+	d.streamMutex.Unlock()
+	return err
+}
+
+func (d *Destination) recvAckWithRetries(ctx context.Context) (*pb.Ack, error) {
+	d.streamMutex.Lock()
+	if d.stream == nil {
+		err := <-d.errCh
+		return nil, err
+	}
+	ack, err := d.stream.Recv()
+	if err == io.EOF || status.Code(err) == codes.Unavailable {
+		d.streamMutex.Unlock()
+		err := d.waitForReadyState(ctx, d.conn.GetState())
+		if err != nil {
+			return nil, err
+		}
+		return d.recvAckWithRetries(ctx)
+	}
+	d.streamMutex.Unlock()
+	return ack, err
+}
+
+// waitForReadyState waits for the connection state to change into the ready state, returns an error if `maxDowntime` is reached.
+func (d *Destination) waitForReadyState(ctx context.Context, currentState connectivity.State) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, d.config.MaxDowntime)
+	defer cancel()
+	// loop until state is Ready, or timeout is reached
+	for {
+		// wait for the state to change
+		ok := d.conn.WaitForStateChange(timeoutCtx, currentState)
+		if !ok {
+			return fmt.Errorf("maxDowntime time reached while waiting for server to reconnect")
+		}
+		currentState = d.conn.GetState()
+		// break loop if state is Ready
+		if currentState == connectivity.Ready {
+			break
+		}
+		select {
+		case err := <-d.errCh:
+			return err
```

Comment:
since we're now listening to `<-d.t.Dying()` and open `ctx` is attached to the tomb, then we can get rid of the case listening to `<-d.openCtx.Done()` here right? or even just keep one of them

Reply:
Right, that should work.